File: SWEA/D3/2806. N－Queen/N－Queen.py
def check(r, c):
    if diagonal1[r + c]:
        return False
    if diagonal2[r - c + n - 1]:
        return False
    if column[c]:
        return False
    return True


def n_queen(r):
    global ans
    if r == n:
        ans += 1
        return
    for c in range(n):
        if check(r, c):
            diagonal1[r + c] = True
            diagonal2[r - c + n - 1] = True
            column[c] = True
            n_queen(r + 1)
            diagonal1[r + c] = False
            diagonal2[r - c + n - 1] = False
            column[c] = False


t = int(input())
for case in range(1, t + 1):
    n = int(input())
    diagonal1 = [False] * (2 * n - 1)
    diagonal2 = [False] * (2 * n - 1)
    column = [False] * n
    ans = 0
    n_queen(0)
    print(f'#{case} {ans}')

File: SWEA/D3/2806. N－Queen/N－Queen.java
import java.io.*;

public class Solution {
	static int N;
	static int ans;
	static boolean[] check_col;
	static boolean[] check1;
	static boolean[] check2;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();
		for (int tc = 1; tc <= T; tc++) {
			N = Integer.parseInt(br.readLine());
			check_col = new boolean[N];
			check1 = new boolean[2 * N - 1];
			check2 = new boolean[2 * N - 1];
			ans = 0;
			n_queen(0);
			sb.append("#").append(tc).append(" ").append(ans).append("\n");
		}
		System.out.print(sb);
	}

	static void n_queen(int row) {
		if (row == N) {
			ans++;
			return;
		}
		for (int col = 0; col < N; col++) {
			if (!check_col[col] && !check1[row + col] && !check2[col - row + (N - 1)]) {
				check_col[col] = true;
				check1[row + col] = true;
				check2[col - row + (N - 1)] = true;
				n_queen(row + 1);
				check_col[col] = false;
				check1[row + col] = false;
				check2[col - row + (N - 1)] = false;
			}
		}
	}
}